Add tokenizer support for URLPattern API

jelee53 opened the pull request:
#### bd6d30811d4248b258082ae01ef4626866f15b40
<pre>
Add tokenizer support for URLPattern API
<a href="https://bugs.webkit.org/show_bug.cgi?id=281730">https://bugs.webkit.org/show_bug.cgi?id=281730</a>
<a href="https://rdar.apple.com/138171576">rdar://138171576</a>

Reviewed by Chris Dumez.

URLPattern API spec for tokenizer: <a href="https://urlpattern.spec.whatwg.org/#tokenizer">https://urlpattern.spec.whatwg.org/#tokenizer</a>

* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/url-pattern/URLPatternTokenizer.cpp: Added.
(WebCore::URLPatternUtilities::isValidNameCodepoint):
(WebCore::URLPatternUtilities::Token::isNull const):
(WebCore::URLPatternUtilities::Tokenizer::getNextCodePoint):
(WebCore::URLPatternUtilities::Tokenizer::seekNextCodePoint):
(WebCore::URLPatternUtilities::Tokenizer::addToken):
(WebCore::URLPatternUtilities::Tokenizer::processTokenizingError):
(WebCore::URLPatternUtilities::Tokenizer::Tokenizer):
(WebCore::URLPatternUtilities::Tokenizer::tokenize):
* Source/WebCore/Modules/url-pattern/URLPatternTokenizer.h: Added.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/285699@main">https://commits.webkit.org/285699@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b22db86a29ff89c5fdc5b4c566699a4b7dd86775

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/73541 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/52970 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/26349 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/77827 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/24779 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/75656 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/62101 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/755 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/57803 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/16221 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/76608 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/47910 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/63276 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/38214 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/44516 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/20762 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/23111 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/66290 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/21111 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/79428 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/857 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/338 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/66186 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/998 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/63288 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/65466 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16180 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/9322 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/7498 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/820 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/3569 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/850 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/836 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/856 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternTokenizer.cpp:
```diff
@@ -0,0 +1,262 @@
+
+bool Token::isNull()
+{
+    return (this->type == TokenType::Open && !this->index && this->value.isEmpty());
```
Please omit the parentheses.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternTokenizer.cpp:
```diff
@@ -0,0 +1,262 @@
+    return u_isIDPart(codepoint);
+}
+
+bool Token::isNull()
```
Can be marked as `const`

cdumez replied:
Not addressed?

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternTokenizer.cpp:
```diff
@@ -0,0 +1,262 @@
+// https://urlpattern.spec.whatwg.org/#add-a-token-with-default-length
+void Tokenizer::addToken(TokenType currentType, int nextPosition, int valuePosition)
+{
+    addToken(currentType, nextPosition, valuePosition, (nextPosition - valuePosition));
```
Please omit parentheses around `nextPosition - valuePosition`

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternTokenizer.h:
```diff
@@ -0,0 +1,64 @@
+    ExceptionOr<Vector<Token>> tokenize(StringView patternStringInput, TokenizePolicy behavior);
+
+private:
+    int m_index { 0 };
```
Any reason we're using `int` instead of `size_t` for all these indexes?

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternTokenizer.h:
```diff
@@ -0,0 +1,64 @@
+
+class Tokenizer {
+public:
+    StringView input;
```
We shouldn't have a class with public data members. Either it's a `struct` or the data members are private with `m_` prefix and we have getter / setters.

cdumez replied:
In this instance, I suspect we want to make the data members private and add a constructor.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternTokenizer.cpp:
```diff
@@ -0,0 +1,262 @@
+
+    ExceptionOr<void> maybeException;
+
+    while (static_cast<unsigned>(this->m_index) < patternStringInput.length()) {
```
With proper typing for m_index, we wouldn't need a cast here.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternTokenizer.h:
```diff
@@ -0,0 +1,64 @@
+
+namespace WebCore {
+
+enum class TokenType : uint8_t { Open, Close, Regexp, Name, Char, EscapedChar, OtherModifier, Asterisk, End, InvalidChar };
```
`TokenType` is too generic a name to be at the global WebCore scope. Same comment for the other types below. I suspect we move all under a new `WebCore::URLPattern` namespace.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternTokenizer.h:
```diff
@@ -0,0 +1,64 @@
+namespace WebCore {
+
+enum class TokenType : uint8_t { Open, Close, Regexp, Name, Char, EscapedChar, OtherModifier, Asterisk, End, InvalidChar };
+enum class TokenizePolicy : uint8_t { Strict, Lenient };
```
`: bool`

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternTokenizer.cpp:
```diff
@@ -0,0 +1,262 @@
+// https://urlpattern.spec.whatwg.org/#get-the-next-code-point
+void Tokenizer::getNextCodePoint()
+{
+    this->m_codepoint = this->input.characterAt(this->m_nextIndex);
```
Please avoid `this->` in your code as much as possible.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternTokenizer.cpp:
```diff
@@ -0,0 +1,262 @@
+void Tokenizer::getNextCodePoint()
+{
+    this->m_codepoint = this->input.characterAt(this->m_nextIndex);
+    this->m_nextIndex++;
```
`++m_nextIndex;`

cdumez replied:
But actually, this doesn't look correct. If the spec says to get the next code point then just incrementing an index like this is not correct. This will just get you the next character, not the next code point.

code points are 32 bits and you can take a look at `StringView::CodePoints::Iterator::operator++()` to see how we iterate code points.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternTokenizer.cpp:
```diff
@@ -0,0 +1,275 @@
+namespace WebCore::URLPatternUtilities {
+
+// https://urlpattern.spec.whatwg.org/#is-a-valid-name-code-point
+static bool isValidNameCodepoint(const UChar codepoint, const bool first)
```
No need for `const` for the parameters. We usually omit it in WebKit unless the parameters are references / pointers.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternTokenizer.h:
```diff
@@ -0,0 +1,66 @@
+
+#pragma once
+
+namespace WebCore::URLPatternUtilities {
```
Does this work? I expected:
```
namespace WebCore {
namespace URLPatternUtilities {
```

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternTokenizer.cpp:
```diff
@@ -0,0 +1,275 @@
+void Tokenizer::getNextCodePoint()
+{
+    if (m_input.is8Bit()) {
+        const LChar* tempLcharPtr = static_cast<const LChar*>(m_input.span8().data());
```
I don't think you need the static_cast here.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternTokenizer.cpp:
```diff
@@ -0,0 +1,275 @@
+        const LChar* tempLcharPtr = static_cast<const LChar*>(m_input.span8().data());
+        U8_NEXT_OR_FFFD(tempLcharPtr, m_nextIndex, m_input.length(), m_codepoint);
+    } else {
+        const UChar* tempUcharPtr = static_cast<const UChar*>(m_input.span16().data());
```
I don't think you need the static_cast here.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternTokenizer.cpp:
```diff
@@ -0,0 +1,275 @@
+{
+    if (m_input.is8Bit()) {
+        const LChar* tempLcharPtr = static_cast<const LChar*>(m_input.span8().data());
+        U8_NEXT_OR_FFFD(tempLcharPtr, m_nextIndex, m_input.length(), m_codepoint);
```
I looked at `StringView::CodePoints::Iterator::operator++()` and it only does a simple pointer increment in the 8-bit case. Why the difference?

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternTokenizer.cpp:
```diff
@@ -0,0 +1,275 @@
+}
+
+// https://urlpattern.spec.whatwg.org/#seek-and-get-the-next-code-point
+void Tokenizer::seekNextCodePoint(int index)
```
Still a lot of `int` instead of `size_t` in this patch. Why does it need to be int?

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternTokenizer.cpp:
```diff
@@ -0,0 +1,270 @@
+// https://urlpattern.spec.whatwg.org/#get-the-next-code-point
+void Tokenizer::getNextCodePoint()
+{
+    if (m_input.is8Bit())
```
You'll need to do something like this to fix the build:
```
if (m_input.is8Bit()) {
    auto characters = m_input.span8();
    U8_NEXT_OR_FFFD(characters, m_nextIndex, m_input.length(), m_codepoint);
} else {
    auto characters = m_input.span16();
    U16_NEXT_OR_FFFD(characters, m_nextIndex, m_input.length(), m_codepoint);
}
```

cdumez replied:
The local variable makes the unused variable error go away.

cdumez replied:
I hit this previously here: https://github.com/WebKit/WebKit/pull/27831/files#r1581848413

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternTokenizer.cpp:
```diff
@@ -0,0 +1,272 @@
+#include "config.h"
+#include "URLPatternTokenizer.h"
+
+#import <unicode/utf16.h>
```
`#include`

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternTokenizer.cpp:
```diff
@@ -0,0 +1,272 @@
+#include "URLPatternTokenizer.h"
+
+#import <unicode/utf16.h>
+#import <unicode/utf8.h>
```
ditto.

cdumez commented on Source/WebCore/Headers.cmake:
```diff
@@ -603,6 +603,7 @@ set(WebCore_PRIVATE_FRAMEWORK_HEADERS
     Modules/url-pattern/URLPatternInit.h
     Modules/url-pattern/URLPatternOptions.h
     Modules/url-pattern/URLPatternResult.h
+    Modules/url-pattern/URLPatternTokenizer.h
```
I doubt this new line is needed, or any of the `url-pattern/` ones above. This is only needed to export headers from WebCore/, so they can be used from WebKit/. None of `url-pattern/` is used from WebKit at the moment.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternTokenizer.cpp:
```diff
@@ -0,0 +1,272 @@
+            int depth = 1;
+            auto regexPosition = m_nextIndex;
+            auto regexStart = regexPosition;
+            bool error = false;
```
Per WebKit coding style (https://webkit.org/code-style-guidelines/#names-bool), boolean variables need a prefix. Here maybe `hasError`

jelee53 committed:
Add tokenizer support for URLPattern API
https://bugs.webkit.org/show_bug.cgi?id=281730
rdar://138171576

Reviewed by Chris Dumez.

URLPattern API spec for tokenizer: https://urlpattern.spec.whatwg.org/#tokenizer

* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/url-pattern/URLPatternTokenizer.cpp: Added.
(WebCore::URLPatternUtilities::isValidNameCodepoint):
(WebCore::URLPatternUtilities::Token::isNull const):
(WebCore::URLPatternUtilities::Tokenizer::getNextCodePoint):
(WebCore::URLPatternUtilities::Tokenizer::seekNextCodePoint):
(WebCore::URLPatternUtilities::Tokenizer::addToken):
(WebCore::URLPatternUtilities::Tokenizer::processTokenizingError):
(WebCore::URLPatternUtilities::Tokenizer::Tokenizer):
(WebCore::URLPatternUtilities::Tokenizer::tokenize):
* Source/WebCore/Modules/url-pattern/URLPatternTokenizer.h: Added.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/285699@main
